enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If an operator does not opt-in to "OLM Supported Conditions", OLM falls on existing behavior.
+- If an operator opts-in to the `NotInterruptible` "OLM Supported Conditions":
+  - Upgrades are not blocked if no CRs exist.
+  - Upgrades are not blocked if a CR exists but the condition is missing
+  - Upgrades are not blocked if a CR exists, the mapped condition type is present and is not equal to `true`.
+  - Upgrades are blocked if a CR exists, the mapped condition type is present and `true`.
+
+
+### Risks and Mitigations
+
+- OLM cannot assume that a Condition on a CR with the type `NotInterruptible` maps to the `NotInterruptible` "OLM Supported Condition". It is impossible for OLM to guarantee that it will not create an "OLM Supported Condition" that is not already being used as a condition type by all operators in the world. By making this feature opt-in only, OLM avoids this conflict.
+- Operators may opt-in to this feature but never update the endpoint as "upgradeable". In these situations, cluster admins must uninstall the operator and upgrade the operator manually.
+
+## Alternatives
+
+There were a number of viable alternatives that were considered when creating this enhancement.
+
+### Alternative 1: OperatorStatus CRD
+
+In this approach, OLM would introduce a new namespaced [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/) called the OperatorStatus CRD. The OperatorStatus object would have no spec. The OperatorStatus's [Status subresource](https://book-v1.book.kubebuilder.io/basics/status_subresource.html) would include a map of conditions based off the  [Kubernetes APIServiceCondition](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podcondition-v1-core). The Maps key's will be consistent with the "Name" of the condition being communicated and the values.
+
+The OperatorStatus CRD would be used by operator to:
+- Communicate conditions and states that should influence how OLM manages the operator, such as preventing OLM from upgrading the operator.
+- Communicate messages to cluster admins, such as slow off-cluster queries or poorly configured CRs. These communications will not influence OLM behavior.
+
+Similarly, OLM may use the OperatorStatus CRD to:
+- Communicate to cluster admins the state of an operator, such as why an operator is unable to be upgraded.
+- Communicate with the operator about upcoming actions.
+
+
+#### Pros
+
+- A single point of communication between OLM and Operators.
+- Feature is opt-in, operators can choose not to take advantage of this feature.
+- This approach can be expanded to support new communication channels easily.
+- Allows communication of conditions that influence OLM or provide helpful messages to Cluster Admins.
+
+#### Cons
+
+- Operators are required to know when they are managed by OLM and write to an OLM resource.
```

Comment:
I thought that there was a thought, that the above implementation would be aggregated to this object and that an operator could choose to write this object OR OLM would use the above annotations to write out to this object that then would be picked up by the operator API? 

Giving some operators more detailed control rather than having to update every CR seems like a valuable addition.

Reply:
> I thought that there was a thought, that the above implementation would be aggregated to this object and that an operator could choose to write this object OR OLM would use the above annotations to write out to this object that then would be picked up by the operator API?

I believe that your thinking of the `Operator API` that @njhale introduced where the conditions of "operator objects" are aggregated.

> Giving some operators more detailed control rather than having to update every CR seems like a valuable addition.

If an operator does not want to communicate state using multiple CRs they can introduce a single CRD that they write status to.

Reply:
> If an operator does not want to communicate state using multiple CRs they can introduce a single CRD that they write status to.

Reply:
I think we didn’t want to expand the API surface because from the F2F we heard “OLM has too many apis” and “we want to simplify things for users”. But in this case if the trade-off is not worth it, and things get actually more complex without a new API, and we are given license to make a new API, we should do so.

@njhale brought up the point that something like an `OperatorStatus` API overlaps a lot with the Status of the `Operator` API that is being introduced, so it could potentially be confusing to the end-user if we offer them both. If the new `Operator` API becomes writeable that could potentially be a good place for this API to live. 

Another argument against introducing a new API is that it further couples OLM to the operators that it runs -- this goes counter to the ongoing work to remove the dependency on a CSV and have OLM run arbitrary kube objects to ease the burden on operator authors. The operator would write to this API (create a CR) only on OLM-managed clusters -- meaning operator behavior becomes conditional on where its running, which further increases the burden of using OLM and potentially lowers adoption of this new feature